Nanopb Generator as a python package

theunkn0wn1 opened the pull request:
This Pull request allows nanopb's generator to be built into a python package.


#   What functionality it fixes/adds.
### additions
- Adds `poetry` as a development dependency, to build wheels from the `generator` folder.
- `pyproject.toml` project configuration file - describes the package as a whole and how to build it. (sdist)
- `nanopb_generator` script path, allows `generator/nanopb_generator` to be invoked from the CLI from any directory.
```
root@0a250617787d:~# ls
root@0a250617787d:~# nanopb_generator
Usage: nanopb_generator.py [options] file.pb ...

Options:
  -h, --help            show this help message and exit
  -x FILE               Exclude file from generated #include list.
# output truncated for brevity
```
- `protoc-gen-nanopb` script path, allows `protoc` to automatically detect the nanopb plugin. this means the plugin doesn't need to be explicitly declared!
```
root@0a250617787d:~# ls
empty.proto
root@0a250617787d:~# protoc --nanopb_out=. empty.proto
root@0a250617787d:~# ls
empty.pb.c  empty.pb.h  empty.proto
root@0a250617787d:~# cat empty.pb.c
/* Automatically generated nanopb constant definitions */
/* Generated by nanopb-0.4.0-dev */

#include "empty.pb.h"

/* @@protoc_insertion_point(includes) */
#if PB_PROTO_HEADER_VERSION != 40
#error Regenerate this file with the current version of nanopb generator.
#endif

PB_BIND(rover_empty_Empty, rover_empty_Empty, AUTO)



/* @@protoc_insertion_point(eof) */
```
## building the wheel
```
>  poetry build
Building nanopb (0.3.9.4)
 - Building sdist
 - Built nanopb-0.3.9.4.tar.gz

 - Building wheel
 - Built nanopb-0.3.9.4-py2.py3-none-any.whl
```
### fixes
`generator/nanopb_generator.py`
 - adjusted to use a relative import from its package, to make it work with python packaging.
#    How can the problem be reproduced / when would the feature be useful.
This PR allows nanopb, when installed as a python package. to be transparent to the end-user when generating sources.

I, for one, found automating NanoPB in my downstream build process to be somewhat frustrating, due to the requirement of embedding nanopb's git repository / binary distributions. Having an install-able python wheel allows downstream users to avoid this. 

This PR also allows nanopb's generators to target python3 as a side-effect, without effecting the shebang line (see #441 )



## Problems
For this pull request I took a minimalistic approach, only changing the bare-minimum to achieve the feature. 
Presently, the Nanopb generators get installed into the `generators` package, preferably I would like to place things under `nanopb.generators`. Unfortunately It would appear that we would need to move the `/generators/` folder to `/nanopb/generators` to achieve that result.

Oh, and please do review the `pyproject.toml` and make sure I got the author information correct.

---------

Closes: #401 
Relates to: #441

PetteriAimonen commented:
Looks great so far!

pyproject.toml looks mostly correct. As far as I know, protobuf-3.6 is currently new enough for all functionality, so the dependency can be made a bit less strict. Also updating the version number in this file has to be added to `tools/set_version.sh` so that it gets updated automatically.

I could merge this as is, but I think it might be good to figure out some solution for the package name. I wonder if it would be possible to have some kind of build rule that copies the generator directory under another name before building the package?

theunkn0wn1 commented:
It would probably work if the build system did something to the effect of
```bash
cd path/to/project/root
mkdir nanopb
cp generator nanopb
poetry build
```
I will look into this and the `tools/set_version.sh` when I get home from work this evening.

theunkn0wn1 commented:
Adjusted the `tools/set_version.sh` tool to cover the `pyproject.toml` file.

I have modified the pyproject.toml file to create the package `nanopb.generator`, this will require the build system to copy some files around, unless you want to change project structure. 

@PetteriAimonen Would you like me to add a built step to Travis to effect this and build a wheel?
```bash
# assuming executing from VCS root
mkdir nanopb
touch nanopb/__init__.py
cp generator nanopb
poetry build
```

PetteriAimonen commented:
This looks good now, thanks a lot!

Do you happen to know how publishing packages on `pip` works? I guess that would be a good place for easy installation.

theunkn0wn1 commented:
You should just be able to `poetry publish`, see https://python-poetry.org/docs/cli/#publish

Here is the short of it:
If you have not yet you will need to make an account on PyPI, under which you can publish packages.
Pip pulls, by default, from PyPI.

To publish to PyPI, you need to create whats called a `wheel`, Poetry handles this for you (`poetry build`).
You upload this wheel to the correct endpoint and you now have a release! (`poetry publish` handles this for you).

For the long answer: [here is the documentation on PyPi](https://pypi.org/help/#publishing), and is a good read for understanding how packaging in python works at a lower level.

PetteriAimonen commented:
Ok, I think it works now: https://pypi.org/project/nanopb/

I reorganized files a bit, moving them to `extra/poetry` and added build shell script for the copying. Also made a jenkins job that will automatically publish packages for dev & stable branches.

theunkn0wn1 commented:
Pulled down master and ran the build script, Looks like its working as intended with the python package. 👍 

Would it be possible to backport these changes for 0.3x so I can `pip install` in my downstream projects? (either that or release pre-release builds for the platformio integration?)

PetteriAimonen commented:
I think I did the backport already, should be on PyPI also :)

theunkn0wn1 commented:
so you did, i missed that. Thanks! ❤️

theunkn0wn1 commented:
the 3x backport appears to be broken. Il file an issue ticket.
